lib/paypal_express/api.rb

```diff
@@ -59,15 +60,27 @@ def purchase_payment(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb
         # Pass extra parameters for the gateway here
         options = {}
 
-        add_required_options(kb_payment_transaction_id, kb_payment_method_id, context, options)
+        if find_value_from_properties(properties, 'from_hpp')
+          token    = find_value_from_properties(properties, 'token')
+          payer_id = find_value_from_properties(properties, 'payer_id')
+          payer_id = find_payer_api(token, kb_account_id, context.tenant_id, properties_to_hash(properties)) if payer_id.nil?
+          options  = {
+              :token    => token,
+              :payer_id => payer_id
+          }
+          gateway_call_proc = Proc.new do |gateway, linked_transaction, payment_source, amount_in_cents, options|
+            gateway.purchase(amount_in_cents, options)
+          end
+        else
+          gateway_call_proc = Proc.new do |gateway, linked_transaction, payment_source, amount_in_cents, options|
+            # Can't use default implementation: the purchase signature is for one-off payments only
+            gateway.reference_transaction(amount_in_cents, options)
+          end
+        end
 
```

Comment:
Also in that `purchase_payment` method, don't we need to handle the case where the call comes back because we had the flag `: create_pending_payment ` specified in the `build_form_descriptor` method below (and in which case the call should return  `PaymentPluginStatus.PENDING` ?